### ISSUE TITLE
Throttle CADisplayLink in Low Power Mode

### ISSUE THREAD

**Issue by 1ec5** [mapbox/mapbox-gl-native#10021]
When `-[NSProcessInfo isLowPowerModeEnabled]` reports that Low Power Mode is enabled, MGLMapView should throttle down the framerate of its CADisplayLink. This would help to mitigate (but not eliminate) the power issues seen in mapbox/mapbox-navigation-ios#370 for example.

/cc @frederoni @ericrwolfe

**Comment by friedbunny** [comment]
This seems a bit extreme — users don’t expect their maps to become choppy or less performant when they enable Low Power Mode. I’d probably be confused and not make the connection, were this to happen to me as a developer or user.

Apple Maps/MapKit and Google Maps/SDK don’t obviously do this, though Google does offer [`GMSMapView.preferredFrameRate`](https://developers.google.com/maps/documentation/ios-sdk/reference/interface_g_m_s_map_view.html#a213c64583f914709fb880b542bcd0d08) and `kGMSFrameRatePowerSave`. This appears to be a more-or-less directly exposed [`CADisplayLink.preferredFramesPerSecond`](https://developer.apple.com/documentation/quartzcore/cadisplaylink/1648421-preferredframespersecond), with some [interesting options](https://maps-apis.googleblog.com/2016/04/advanced-marker-animations-and-more.html):

>You can now take advantage of a new enum, `GMSFrameRate`, which has three options: `kGMSFrameRatePowerSave` (15fps), `kGMSFrameRateConservative` (30fps), and the default `kGMSFrameRateMaximum` (60fps with 30fps fallback). In order to protect devices that don't handle 60fps well, we do a run-time device model check and the `preferredFrameRate` is silently downgraded to 30fps when necessary on older devices.

I’m not sure if `kGMSFrameRatePowerSave` will automatically throttle, but it’s not the default.

**Comment by 1ec5** [comment]
A preferred frame rate option would address this use case while being more transparent to the developer. It would also be possible to implement a preferred frame rate in mbgl even after we wean MGLMapView off CADisplayLink: #2985. However, as far as I can tell, Core Animation offers no framerate option, so it would be problematic for us to introduce a public framerate option and then migrate to Core Animation for camera transitions: #8176.

**Comment by ericrwolfe** [comment]
> Core Animation offers no framerate option, so it would be problematic for us to introduce a public framerate option and then migrate to Core Animation for camera transitions: #8176.

@1ec5 I talked through this with @d-prukop, and I don't think it'd be as problematic as we're imagining. Core Animation could update values within mbgl independently of mbgl's frame rate. 

So even if, for example, Core Animation spit out property values at ~60fps, mbgl could stay rendering at 20fps.

**Comment by d-prukop** [comment]
I do have concerns around how core animation synchs up with mbgl framerate. Is it possible to have CA telling MBGL to update but then have MBGL update on it's own timeline?

**Comment by 1ec5** [comment]
> So even if, for example, Core Animation spit out property values at ~60fps, mbgl could stay rendering at 20fps.

Yes, and this would already be the case if we hook the framerate property up to the existing CADisplayLink. But that would also tie us more strongly to CADisplayLink.

> Is it possible to have CA telling MBGL to update but then have MBGL update on it's own timeline?

We could have mbgl take over the throttling that CADisplayLink currently does. That would allow us to address the battery performance issues, although we’d need to take a close look at whether it exacerbates the map’s lagging relative to CALayers (for view-backed annotations).

**Comment by d-prukop** [comment]
> whether it exacerbates the map’s lagging relative to CALayers (for view-backed annotations)

This was my main concern. It sounds like we need to test this out. @1ec5 How can we get to the bottom of this?

**Comment by 1ec5** [comment]
https://github.com/mapbox/mapbox-gl-native/issues/2985#issuecomment-294262394 is the latest on the effort to remove CADisplayLink. It’s worth checking it out – maybe we’ve overlooked something.
